### PR TITLE
[RLlib] RLModule API: `SelfSupervisedLossAPI` for RLModules that bring their own loss (algo independent).

### PR DESCRIPTION
**Pull request by sven1977**
RLModule API: `SelfSupervisedLossAPI` for RLModules that bring their own loss (algo independent).
* `Learner` now checks whether any RLModule (in MultiRLModule) implements this API and if yes, calls the Module's own `compute_self_supervised_loss` method (instead of the Learner's `compute_loss_for_module()` method).
* Updated curiosity RLModule to implement this API.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simonsays1980 — APPROVED** [review]
LGTM.

File: rllib/examples/rl_modules/classes/intrinsic_curiosity_model_rlm.py [hunk 1]
```diff
@@ -198,16 +195,17 @@ def _forward_train(self, batch, **kwargs):
...
+        self,
         *,
         learner: "TorchLearner",
         module_id: ModuleID,
         config: "AlgorithmConfig",
         batch: Dict[str, Any],
         fwd_out: Dict[str, Any],
     ) -> Dict[str, Any]:
-        module = learner.module[module_id]
+        module = learner.module[module_id].unwrapped()
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/examples/rl_modules/classes/intrinsic_curiosity_model_rlm.py:208]: correct

File: rllib/algorithms/ppo/ppo_learner.py [hunk 2]
```diff
@@ -106,15 +108,24 @@ def after_gradient_based_update(
...
 
     @abc.abstractmethod
     def _update_module_kl_coeff(
         self,
         *,
         module_id: ModuleID,
         config: PPOConfig,
+        kl_loss: float,
     ) -> None:
         """Dynamically update the KL loss coefficients of each module with.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/ppo/ppo_learner.py:130]: fixed

File: rllib/examples/learners/classes/intrinsic_curiosity_learners.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+from ray.rllib.core import Columns, DEFAULT_MODULE_ID
+from ray.rllib.core.learner.torch.torch_learner import TorchLearner
+from ray.rllib.core.rl_module.rl_module import RLModule
+from ray.rllib.utils.typing import EpisodeType
+
+ICM_MODULE_ID = "_intrinsic_curiosity_model"
+
+
+class DQNTorchLearnerWithCuriosity(DQNRainbowTorchLearner):
+    def build(self) -> None:
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/examples/learners/classes/intrinsic_curiosity_learners.py:27]: We could, but again, this should be done inside the Learner, imo.

But you have a good point: How can we make this even easier for the user? Maybe offer a better way to customize the Learner pipeline? Currently, users can only prepend connector pieces to the beginning, then RLlib adds the default pieces to the end. But here, we need a (custom) connector piece to move all the way to the end, which is not possible with the `config.learner_connector` property.

File: rllib/examples/rl_modules/classes/intrinsic_curiosity_model_rlm.py [hunk 4]
```diff
@@ -198,16 +195,17 @@ def _forward_train(self, batch, **kwargs):
     def get_train_action_dist_cls(self):
         return TorchCategorical
 
-    @staticmethod
-    def compute_loss_for_module(
+    @override(SelfSupervisedLossAPI)
+    def compute_self_supervised_loss(
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/examples/rl_modules/classes/intrinsic_curiosity_model_rlm.py:199]: See my answer to your comment above.

**Pull request merged**